packages/core/src/envelope.ts

```diff
@@ -128,20 +132,22 @@ function createEventEnvelopeHeaders(
     ...(event.type === 'transaction' &&
       event.contexts &&
       event.contexts.trace && {
-        // TODO: Grab this from baggage
         trace: dropUndefinedKeys({
           // Trace context must be defined for transactions
           // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-          trace_id: event.contexts!.trace.trace_id as string,
-          environment: event.environment,
-          release: event.release,
-          transaction: event.transaction,
-          user: event.user && {
-            id: event.user.id,
-            segment: event.user.segment,
-          },
+          trace_id: (event.contexts!.trace as Record<string, unknown>).trace_id as string,
           public_key: dsn.publicKey,
-        }),
+          environment: baggage && baggage.environment,
+          release: baggage && baggage.release,
+          transaction: baggage && baggage.transaction,
+          ...(baggage &&
+            (baggage.userid || baggage.usersegment) && {
+              user: {
+                id: baggage.userid,
+                segment: baggage.usersegment,
+              },
+            }),
+        } as EventTraceContext),
```

Comment:
A slightly shorter way to write this might be to unpack those values right after you pull out `baggage`

```js
const { environment, release, transaction, userid, usersegment } = baggage || {}
```

and then you can just do 

```suggestion
          environment,
          release,
          transaction,
          ...(userid || usersegment) && {
              user: {
                id: userid,
                segment: usersegment,
              },
            }),
        } as EventTraceContext),
```

Reply:
Thanks, that's a lot cleaner! 561d49201239cccff67f6dac4a753ecf517b7eac